### DISCUSSION TITLE
How to recover sub-expressions from an expression?

### DISCUSSION THREAD

**Discussion by chris-hermans** [casadi/casadi#3445] [Q&A] [answered]
I've got a question that I could not find an answer to in the documentation. I'm making the assumption that the expression is stored internally as an expression tree, before it is manipulated to be sent to the solver.

My goal is to manipulate a constraint that I pass to a function. This constraint is going to be an expression of the form `expr1 == expr2` or `expr1 <= expr2`.

Provided I get the parent expression, how can I recover `expr1`, `==`, `expr2` from it?

Thanks in advance,
Chris

**Comment by jgillis** [comment, marked as answer]
You are looking for `dep(i)` and `op()`

**Reply by chris-hermans** [reply]
Thanks, that does seem to do the trick!

**Reply by sandeep026** [reply]
This can be useful for me. I did not find any documentation on the python API. @chris-hermans or developers , can you explain what each method does and how to use them? 
Given, expr1<=expr2, convert to expr1-expr2<=0.
